dartdoc_test/lib/src/reporter.dart

```diff
@@ -0,0 +1,186 @@
+abstract class SourceFileReporter {
+  void reportCodeSample(Uri generatedUrl,
+      [void Function(CodeSampleReporter)? fn]);
+}
+
+abstract class CodeSampleReporter {
+  void reportIssues(Iterable<Issue> issues);
+}
+
+final class _ReporterForStdout extends Reporter
+    implements SourceFileReporter, CodeSampleReporter {
+  final Logger logger;
+
+  _ReporterForStdout({required super.verbose})
+      : logger = Logger(
+          (message, level) {
+            if (verbose || LogLevel.debug != level) {
+              io.stdout.writeln(message);
+            }
+          },
+        );
+
+  @override
+  void reportSourceFile(String filename,
+      [void Function(SourceFileReporter)? fn]) {
+    fn?.call(this);
+    final issues = _issues.where((issue) => issue.path == filename);
+    reportIssues(issues);
+  }
+
+  @override
+  void reportCodeSample(Uri generatedUrl,
+      [void Function(CodeSampleReporter)? fn]) {
+    fn?.call(this);
+    final issues = _issues
+        .where((issue) => issue.generatedSpan?.sourceUrl == generatedUrl);
+    reportIssues(issues);
+  }
+
+  @override
+  void reportIssues(Iterable<Issue> issues) {
+    for (final issue in issues) {
+      _reportIssue(issue);
+    }
+  }
+
+  void _reportIssue(Issue issue) {
+    if (issue.commentSpan == null) {
+      logger.debug(
+        '${issue.generatedSpan?.format(issue.message)} '
+        '(ignored because issue occurs in the generated code)',
+      );
+    } else {
+      logger.debug(
+          'original error: ${issue.generatedSpan?.format(issue.message)}');
+      logger.info('${issue.commentSpan!.info(issue.message)}\n');
+    }
+  }
+}
+
+final class _RepoterForTest extends Reporter
+    implements SourceFileReporter, CodeSampleReporter {
+  _RepoterForTest({required super.verbose});
+
+  @override
+  void reportSourceFile(String filename,
+      [void Function(SourceFileReporter)? fn]) {
+    final path = p.relative(filename, from: currentDir.path);
+    group('[dartdoc_test] $path', () {
+      fn?.call(this);
+      final issues = _issues.where((issue) => issue.path == filename);
+      reportIssues(issues);
+    });
+  }
+
+  @override
+  void reportCodeSample(Uri generatedUrl,
+      [void Function(CodeSampleReporter)? fn]) {
+    final path = p.relative(generatedUrl.path, from: currentDir.path);
+    group('[dartdoc_test] $path', () {
+      fn?.call(this);
+      final issues = _issues
+          .where((issue) => issue.generatedSpan?.sourceUrl == generatedUrl);
+      reportIssues(issues);
+    });
+  }
+
+  @override
+  void reportIssues(Iterable<Issue> issues) {
+    for (final issue in issues) {
+      _reportIssue(issue);
+    }
+  }
+
+  void _reportIssue(Issue issue) {
+    test(
+      'test for ${issue.commentSpan?.text}',
+      () {
+        if (issue.commentSpan != null) {
+          fail(issue.commentSpan!.info(issue.message));
+        }
+      },
+      // skip: !_verbose && issue.generatedSpan == null,
+    );
+  }
+}
+
+class Issue {
```

Comment:
Should this be final? or do you intent for people using the API to make their own subclasses?